File: grid_cap_regress.py
import numpy as np
import matplotlib.pyplot as plt

np.random.seed(98)
m = 500
j = 1j

# injeções nodais iniciais
S = -np.array([0.224, 0.708, 1.572, 0.072]) * np.exp(j * 0.3176)
I = np.conj(S).reshape(-1,1)  # (4,1)

# admitâncias
y12 = 1 - j*10
y13 = 2 * y12
y23 = 3 - j*20
y34 = y23
y45 = 2 * y12

Y_full = np.array([
    [y12+y13,  -y12,     -y13,      0,     0],
    [   -y12, y12+y23,   -y23,      0,     0],
    [   -y13,   -y23, y13+y23+y34, -y34,   0],
    [      0,      0,      -y34, y34+y45, -y45]
], dtype=complex)
Y = Y_full[:4,:4]  # retirar coluna do slack

e4 = np.random.randn(4, m) * 0.25
e1 = np.random.randn(m) * 0.5
i1w = [I[0,0].real]

i12, i13, i23 = [], [], []

# simulação do circuito
for t in range(m-1):
    next_I = 0.65 * I[:, t:t+1] + e4[:, t:t+1]
    i1w.append(0.75 * i1w[-1] + e1[t])
    next_I[0,0] = -i1w[-1] + j * next_I[0,0].imag

    I = np.hstack((I, next_I))
    
    v = 1 + np.linalg.inv(Y) @ I[:, t+1]

    I12 = y12 * (v[0] - v[1])
    I13 = y13 * (v[0] - v[2])
    I23 = y23 * (v[1] - v[2])
    i12.append(np.abs(I12) * np.sign(I12.real))
    i13.append(np.abs(I13) * np.sign(I13.real))
    i23.append(np.abs(I23) * np.sign(I23.real))

Ii = np.abs(I) * np.sign(I.real)  # (4,m)

# plots
plt.figure(figsize=(10,9))

plt.subplot(3,1,1)
plt.plot(I.real.T)
plt.title('Injeções nodais (Parte Real)')
plt.xlabel('Time stamp')
plt.ylabel('I [pu]')
plt.legend([f'Nó {i+1}' for i in range(I.shape[0])])

plt.subplot(3,1,2)
plt.plot(I.imag.T, '--')
plt.title('Injeções nodais (Parte Imaginária)')
plt.xlabel('Time stamp')
plt.ylabel('I [pu]')
plt.legend([f'Nó {i+1}' for i in range(I.shape[0])])

plt.subplot(3,1,3)
lines = np.vstack([i12, i13, i23]).T
plt.plot(lines)
plt.title('Correntes nos ramos')
plt.xlabel('Time stamp')
plt.ylabel('I_branch [pu]')
plt.legend(['12','13','23'])

plt.tight_layout()
plt.show()

# detetar violações e OLS
rate_12, rate_13, rate_23 = 0.5 , 1.0, 0.25
ij = lines
X, y = [], []
for t in range(len(ij)):
    pos_vio = max(ij[t,0]-rate_12, ij[t,1]-rate_13, ij[t,2]-rate_23)
    if pos_vio > 0:
        X.append(Ii[:,t])
        y.append(pos_vio)
X = np.column_stack([X, np.ones(len(X))])
y = np.array(y).reshape(-1,1)

beta = np.linalg.inv(X.T @ X) @ X.T @ y
res = y - X @ beta

print("Coeficientes estimados:", beta.flatten())
print("Erro médio (norm/k):", np.linalg.norm(res)/len(y))

File: Grid_Cap_Regress.m
% hosting capacity maximization (kite mesh)
clc; clear all

% power injections (negative load)
S=-[0.224    0.708    1.572    0.072]*exp(j*.3176); % acos(0.95) ind
I=conj(S).';

% line admittances of the kite grid 
y12=1-j*10;
y13=2*y12;
y23=3-j*20;
y34=y23;
y45=2*y12;

% kite admittance matrix (slack on 5)
Y=[y12+y13 -y12     -y13          0   0
  -y12      y12+y23 -y23          0   0
  -y13     -y23      y13+y23+y34 -y34 0
   0        0       -y34          y34+y45  -y45];

% AR rnd process for injections
i1w(1)=real(I(1)); % in node 1
m=500; rng(98);
e4=randn(4,m)*.25;
e1=randn(1,m)*.5;
for t=1:m-1
    I(:,t+1)=0.65*I(:,t)+e4(:,t);  % all other nodes
    i1w(t+1)=0.75*i1w(t)+e1(t);
    I(1,t+1)=-i1w(t+1)+j*imag(I(1,t+1)); % (-)i1w from active load to inj.

    v=1+Y(1:4,1:4)^-1*I(:,t+1); % ohm & kirchhoff's circuit laws
    I12=y12*(v(1)-v(2)); i12(t+1)=abs(I12)*sign(real(I12)); 
    I13=y13*(v(1)-v(3)); i13(t+1)=abs(I13)*sign(real(I13));
    I23=y23*(v(2)-v(3)); i23(t+1)=abs(I23)*sign(real(I23));
end
Ii=abs(I).*sign(real(I));  % abs & sign to repliclate meter RMS info

title('input (plot 1) / output (plot 2)') 
subplot(2,1,1),  plot(1:m,I)
xlabel('Time stamp')
ylabel('Nodal injections') 

subplot(2,1,2),  plot(1:m,([i12; i13; i23]))
xlabel('Time stamp')
ylabel('Line currents') 

rate_12=inf; %.25; 
rate_13=1; 
rate_23=inf; %.50
ij=[i12' i13' i23'];
Ii=Ii';
k=0;
for t=1:m
    pos_vio=max([ ij(t,1)-rate_12  ij(t,2)-rate_13  ij(t,3)-rate_23]);
    neg_vio=max([-ij(t,1)-rate_12 -ij(t,2)-rate_13 -ij(t,3)-rate_23]);
    if  pos_vio > 0
        k=k+1; y(k,:)=pos_vio; X(k,1:4)=Ii(t,1:4);
    end
%     if  neg_vio > 0 
%         k=k+1; y(k,:)=neg_vio; X(k,1:4)=-Ii(t,1:4);
%     end
end

X=[X ones(k,1)];
beta =(X'*X)^-1*X'*y

res=y-X*beta;
norm(res)/k